classes/Wrapper/ProductWrapper.php

```diff
@@ -20,97 +20,237 @@
 
 namespace PrestaShop\Module\Ps_Googleanalytics\Wrapper;
 
+use Configuration;
 use Context;
-use Currency;
-use PrestaShop\Module\Ps_Googleanalytics\Hooks\WrapperInterface;
-use Product;
-use Tools;
+use Db;
+use Manufacturer;
+use PrestaShop\PrestaShop\Adapter\Presenter\Product\ProductLazyArray;
+use PrestaShop\PrestaShop\Adapter\Presenter\Product\ProductListingLazyArray;
+use Shop;
 
-class ProductWrapper implements WrapperInterface
+class ProductWrapper
 {
     private $context;
+    private $categories = [];
+    private $homeCategory = 0;
 
     public function __construct(Context $context)
     {
         $this->context = $context;
     }
 
     /**
-     * wrap products to provide a standard products information for google analytics script
+     * Takes provided list of product (lazy) arrays and converts it to a format that GA4 requires.
+     *
+     * @param array $productList
+     * @param bool $useProvidedQuantity Should provided quantity be used, usually for cart related events
+     *
+     * @return array Item data standardized for GA
      */
-    public function wrapProductList($products)
+    public function prepareItemListFromProductList($productList, $useProvidedQuantity = false)
     {
-        $result_products = [];
-        if (!is_array($products)) {
-            return;
-        }
+        $items = [];
 
-        $currency = new Currency($this->context->currency->id);
-        $usetax = (Product::getTaxCalculationMethod((int) $this->context->customer->id) != PS_TAX_EXC);
+        // Check we actually got some product
+        if (empty($productList)) {
+            return [];
+        }
 
-        foreach ($products as $index => $product) {
-            if ($product instanceof Product) {
-                $product = (array) $product;
+        // Preload categories for the whole list
+        $productIds = [];
+        foreach ($productList as $product) {
+            if (!empty($product['id_product'])) {
+                $productIds[] = $product['id_product'];
+            } elseif (!empty($product['id'])) {
+                $productIds[] = $product['id'];
             }
+        }
+        $this->loadCategories($productIds);
 
-            if (!isset($product['price'])) {
-                $product['price'] = (float) Tools::displayPrice(Product::getPriceStatic((int) $product['id_product'], $usetax), $currency);
-            }
-            $result_products[] = $this->wrapProduct($product, [], $index);
+        // Prepare each item and override the counter
```

Comment:
I'm not sure to understand why you need to override the counter

Reply:
Because I render `index: 0` when doing a single product.
When rendering a list, I will do 0,1,2... etc.